codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/QueryShapeSerVisitor.java

```diff
@@ -0,0 +1,214 @@
+        writer.write("const entries: any = {};");
+        // Set up a counter to increment the member entries.
+        writer.write("let counter = 1;");
+        // Use the keys as an iteration point to dispatch to the input value providers.
+        writer.openBlock("Object.keys(input).forEach(key => {", "});", () -> {
+            // Prepare the key's entry.
+            // Use the @xmlName trait if present on the member, otherwise use "key".
+            MemberShape keyMember = shape.getKey();
+            Shape keyTarget = model.expectShape(keyMember.getTarget());
+            String keyName = keyMember.getTrait(XmlNameTrait.class)
+                    .map(XmlNameTrait::getValue)
+                    .orElse("key");
+            writer.write("const keyLoc: string = \"entry.$L.\" + counter;", keyName);
+            writer.write("entries[keyLoc] = $L;", keyTarget.accept(getMemberVisitor("key")));
+
+
+            // Prepare the value's entry.
+            // Use the @xmlName trait if present on the member, otherwise use "value".
+            MemberShape valueMember = shape.getValue();
+            Shape valueTarget = model.expectShape(valueMember.getTarget());
+            String valueName = valueMember.getTrait(XmlNameTrait.class)
+                    .map(XmlNameTrait::getValue)
+                    .orElse("value");
+            writer.write("const valueLoc: string = \"entry.$L.\" + counter;", valueName);
+            writer.write("entries[valueLoc] = $L;", valueTarget.accept(getMemberVisitor("input[key]")));
+
+            writer.write("counter++;");
+        });
+
+        writer.write("return entries;");
+    }
+
+    @Override
+    protected void serializeStructure(GenerationContext context, StructureShape shape) {
+        TypeScriptWriter writer = context.getWriter();
+
+        // Set up a location to store all of the entry pairs.
+        writer.write("const entries: any = {};");
+
+        // Serialize every member of the structure if present.
+        shape.getAllMembers().forEach((memberName, memberShape) -> {
+            String inputLocation = "input." + memberName;
+            writer.openBlock("if ($L !== undefined) {", "}", inputLocation,
+                    () -> serializeNamedMember(context, memberName, memberShape, inputLocation));
+        });
+
+        writer.write("return entries");
+    }
+
+    private void serializeNamedMember(
+            GenerationContext context,
+            String memberName,
+            MemberShape memberShape,
+            String inputLocation
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        // Grab the target shape so we can use a member serializer on it.
+        Shape target = context.getModel().expectShape(memberShape.getTarget());
+
+        // Use the @xmlName trait if present on the member, otherwise use the member name.
+        String locationName = memberShape.getTrait(XmlNameTrait.class)
+                .map(XmlNameTrait::getValue)
```

Comment:
You need to use the `Ec2QueryName` trait here too.

Reply:
Support for that trait will come along with the support for the `aws.ec2` protocol.